Force quiet on all RunnerClients made by the salt-master

mbirtwell opened the pull request:
### What does this PR do?
Ensure that quiet is set when creating RunnerClient objects inside the salt master so that it doesn't spend time writing the results of the runner command to stdout (which probably is /dev/null). This can be significant when the command is jobs.list_job on a large highstate.

### What issues does this PR fix or reference?
Fixes: #58203

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

Commit:
Force quiet on all RunnerClients made by the salt-master

Fixes #58203

waynew commented on salt/master.py:
```diff
@@ -103,6 +103,12 @@
 
+def _make_master_daemon_runner_client(opts):
+    ropts = dict(opts)
+    ropts["quiet"] = True
```
I feel like this should be `ropts.setdefault("quiet", True)`, otherwise it's going to override the option. Unless that's the intention 🤔

mbirtwell replied:
Certainly that was the behaviour of the code that I was inspired by (line 198), but I don't know if that was the intention of that code. I thought it best to be consistent in the new place that I was applying this to. But if we're happy to change the behaviour for that bit. Then what you suggest works for me. Or maybe there's a reason that's not visible to me that means that we actually want different behaviour in these two places.

waynew replied:
Well, we're on the same page there, then 😂 I'm inclined towards consistency, I'd just like to know the original /why/, to know whether or not it's a foolish consistency 😝

mbirtwell replied:
I'm going to take a punt at some logic to support the way that this was and therefore keeping this the same and making the other place where we create a RunnerClient consistent with the first:

The quiet flag seems only to be used to control the main human facing output for the RunnerClient this is the result of the salt runner that you are executing. It obviously makes sense to display this when the RunnerClient is being used from inside salt-run. However when the RunnerClient is being used from with in the salt-master it never make sense to display this because the salt-master is acting as a server and what ever client initiated the request is responsible for displaying this to the user. Therefore the RunnerClient should always be used with quiet=False in the salt-master.

waynew reviewed (changes requested):
Hey @mbirtwell thanks for the PR - we'll definitely need some tests for this, though I expect this should also get the rest of the @saltstack/team-core to take a look.

mbirtwell commented:
I'm not sure how to go about testing this. In tests/unit/test_master.py there are several tests for `salt.master.ClearFuncs.runner` but none of them actually execute the runner function. They all test `salt.master.ClearFuncs.runner` returning early with some kind of error (e.g. authorisation) before it gets to execute the runner requested. `salt.master.ClearFuncs.runner` would then generate a sub process to actually execute the runner function. It's this subprocess that I've changed the behaviour of. 

Maybe an integration test is more appropriate. I can't find anything that seems to invoke the salt master's runner interface directly. There is `tests.integration.netapi.test_client.NetapiClientTest.test_runner` which does so through the net api, which incidentally is relevant to my use case because I've come at this from the salt api. But it's very indirect at this point. And that test is already marked as skipped because there's something else broken that prevents it from actually testing anything (according to the comment).

Whilst looking for testing examples I came across `salt.daemons.masterapi.LocalFuncs.runner`, which looks very similar to the function that I was modifying (`salt.master.ClearFuncs.runner`). The only usage I can find (other than tests) of  `salt.daemons.masterapi.LocalFuncs` is to construct the `masterapi` attribute of `salt.master.ClearFuncs`, which in turn seems to be unused.

Commit:
Pre-commit hook changes

Commit:
Merge branch 'master' into set_quiet_flag_for_runner_client

sagetherage commented:
@mbirtwell looks like this has a merge conflict to address and it still needs a test case, please. : )

Commit:
Merge branch 'master' of https://github.com/saltstack/salt into set_quiet_flag_for_runner_client

mbirtwell commented:
@sagetherage I've updated to fix the merge conflict. I previously wrote a bunch of stuff about testing this and would like a response to that before I try and tackle it. In summary I can't see a good example of testing this kind of change in the current test suite. Especially given that most of the change in behaviour occurs in a different process to one executing the code that I've changed. Please could you point me in the right direction.

sagetherage commented:
@mbirtwell A great way to tackle tests that are particularly complicated or like this one, where it affects another part of the code base, is to attend one of the test clinics during the week. You can subscribe or see them on the public [Salt Community Events Calendar](https://calendar.google.com/calendar/u/0?cid=c2FsdHN0YWNrLmNvbV9tZDczYzN1ZmNzMmVxYnNtbW5pa2U0ZW04MEBncm91cC5jYWxlbmRhci5nb29nbGUuY29t) and if that does not work with your schedule @waynew or @garethgreenaway can be reached on IRC or Slack for direct messages about this PR, as well. I hope that helps.

Also, this PR is failing on pre-commit, we require pre-commit to be run prior, but many PRs exists prior to the requirement -- so to fix that after the PR is created youcan try to pull down the PR locally, make a change (such as adding at test), stage the changed files, and then run the pre-commit manually:
```
    pre-commit run isort - run all the hooks
    pre-commit run black - run black
```
the force a push. This has worked for others, but not always, so we are working on globally addressing this, but in the mean time if you get help with tests also ask about pre-commit. Let me know if I can help more, happy to.

waynew reviewed (approved):
Yeah, I'm not sure how you could really test the particular places that you're changing. You *could*... do something. Lol. I mean, you can test your single function that you're creating.

One other approach that might work, but might be a pain to setup - you could add an integration test to the pytest suite, and use `capsys`, the pytest fixture to capture stdout/err. Then you just ensure that the thing is configured to use stdout/err and then just ensure that it doesn't dump whatever messages to out/err 🤷 

But... that's somewhat complicated and I'm leaning to recommending a 3-approval override. But if you *do* implement that test then 🚀  and we don't need to worry about that! 😁

mbirtwell commented:
I stand by this PR, my former employer had been using this in production for more than a year when I left, I believe that the check failures were flakes and not to do with the changes in this PR and I don't really see this a sensibly testable. But I want to close this to get it off my dashboard.
